rct_optimizations/src/rct_optimizations/validation/camera_intrinsic_calibration_validation.cpp

```diff
@@ -0,0 +1,120 @@
+#include <rct_optimizations/validation/camera_intrinsic_calibration_validation.h>
+#include <boost/accumulators/accumulators.hpp>
+#include <boost/accumulators/statistics.hpp>
+
+namespace rct_optimizations
+{
+VirtualCorrespondenceResult measureVirtualTargetDiff(const Correspondence2D3D::Set &correspondences,
+                                                 const CameraIntrinsics &intr,
+                                                 const Eigen::Isometry3d &camera_to_target_guess,
+                                                 const double pnp_sq_error_threshold)
+{
+  // Create a lambda for doing the PnP optimization
+  auto solve_pnp = [&intr, &camera_to_target_guess, &pnp_sq_error_threshold](
+                     const Correspondence2D3D::Set &corr) -> Eigen::Isometry3d {
+    // Create the first virtual target PnP problem
+    PnPProblem problem;
+    problem.intr = intr;
+    problem.correspondences = corr;
+    problem.camera_to_target_guess = camera_to_target_guess;
+
+    PnPResult result = optimize(problem);
+    if (!result.converged || result.final_cost_per_obs > pnp_sq_error_threshold)
+    {
+      std::stringstream ss;
+      ss << "PnP optimization " << (result.converged ? "converged" : "did not converge")
+         << " with residual error of " << result.final_cost_per_obs << " ("
+         << pnp_sq_error_threshold << " max)";
+      throw std::runtime_error(ss.str());
```

Comment:
Might be a good place to throw our [`OptimizationException`](https://github.com/Jmeyer1292/robot_cal_tools/blob/bea5fda6ee4a3bdd3f7ea3fcb93a329f88364432/rct_optimizations/include/rct_optimizations/types.h#L128-L132).